src/posix/platform/netif.cpp

```diff
@@ -384,6 +427,122 @@ static void processNetifEvent(otInstance *aInstance)
+    }
+}
+
+static bool isMulticastAddressFiltered(const struct in6_addr &aAddress)
+{
+    bool found = false;
+
+    for (size_t i = 0; i < sizeof(kFilterMulticastAddresses) / sizeof(kFilterMulticastAddresses[0]); i++)
+    {
+        if (memcmp(&aAddress, &kFilterMulticastAddresses[i].mFields.m8, sizeof(aAddress)) == 0)
+        {
+            found = true;
+            break;
+        }
+    }
+
+    return found;
+}
+
+static void processMLDEvent(otInstance *aInstance)
+{
+    const size_t        kMaxMLDEvent = 8192;
+    uint8_t             buffer[kMaxMLDEvent];
+    ssize_t             bufferLen = -1;
+    struct sockaddr_in6 srcAddr;
+    socklen_t           addrLen;
+    bool                fromSelf = false;
+    MLDv2Header *       hdr      = reinterpret_cast<MLDv2Header *>(buffer);
+    size_t              offset;
+    uint8_t             type;
+    struct ifaddrs *    ifAddrs = NULL;
+
+    bufferLen = recvfrom(sMLDMonitorFd, buffer, sizeof(buffer), 0, reinterpret_cast<sockaddr *>(&srcAddr), &addrLen);
+    VerifyOrExit(bufferLen > 0);
+
+    type = buffer[0];
+    VerifyOrExit(type == kICMPv6MLDv2Type && bufferLen >= static_cast<ssize_t>(sizeof(MLDv2Header)));
+
+    // Check whether it is sent by self
+    VerifyOrExit(getifaddrs(&ifAddrs) == 0);
+    for (struct ifaddrs *ifAddr = ifAddrs; ifAddr != NULL; ifAddr = ifAddr->ifa_next)
+    {
+        if (ifAddr->ifa_addr != NULL && ifAddr->ifa_addr->sa_family == AF_INET6 &&
+            strncmp(sTunName, ifAddr->ifa_name, IFNAMSIZ) == 0)
+        {
+            struct sockaddr_in6 *addr6 = reinterpret_cast<struct sockaddr_in6 *>(ifAddr->ifa_addr);
+
+            if (memcmp(&addr6->sin6_addr, &srcAddr.sin6_addr, sizeof(in6_addr)) == 0)
+            {
+                fromSelf = true;
+                break;
+            }
+        }
+    }
+    VerifyOrExit(fromSelf);
+
+    hdr    = reinterpret_cast<MLDv2Header *>(buffer);
+    offset = sizeof(MLDv2Header);
+
+    for (size_t i = 0; i < hdr->mNumRecords && offset < static_cast<size_t>(bufferLen); i++)
+    {
+        if (static_cast<size_t>(bufferLen) >= (sizeof(MLDv2Record) + offset))
+        {
+            MLDv2Record *record = reinterpret_cast<MLDv2Record *>(&buffer[offset]);
+
+            if (!isMulticastAddressFiltered(record->mMulticastAddress))
+            {
+                otIp6Address address;
+
+                memcpy(&address.mFields.m8, &record->mMulticastAddress, sizeof(address.mFields.m8));
+                if (record->mRecordType == kICMPv6MLDv2RecordChangeToIncludeType)
+                {
+                    otIp6SubscribeMulticastAddress(aInstance, &address);
+                }
+                else if (record->mRecordType == kICMPv6MLDv2RecordChangeToExcludeType)
+                {
+                    otIp6UnsubscribeMulticastAddress(aInstance, &address);
+                }
+            }
+
+            offset += sizeof(MLDv2Record) + sizeof(in6_addr) * record->mNumSources;
```

Comment:
Do we need to care about the encoding (little/big endian) of the `mNumSources` field? 
We cast a received buffer pointer to `MLDv2Record` and then read the field as `uint16_t` (I guess the complier/platform endian-ness would come into picture when we read it).

Reply:
You are right, we need to add ntohs. It was not causing a crash because of the data length guard.